### PR TITLE
Custom labels for readability and CLI output

### PR DESCRIPTION
**Pull request by Fishbowler**
## Proposed Changes

Builds on the work of @SuperRoach in #1238

Adds a `label` attribute to a number of commands. This:
* adds readability to the file as to the purpose of a particular step (especially relevant for tapping on points or vague element selectors) 
* is also returned in the `description()` method of those commands to be displayed in the CLI output.

I'm a first-timer, so have some outstanding questions:
* Does this need a Docs PR ready for merge before this can be merged?
* Is there additional testing that could/should be done?
* Is partial support for `label` acceptable, so long as it's documented?

## Testing

* Extended Unit tests
* Added new Integration test
* Tested locally by amending my own flows (the reason I'm working on this 😉) 

## Issues Fixed

Solves #1229 and provides a workaround for #1226

### CODE REVIEW

**Comment by DannyCork** [comment]
Can we get this reviewed please? Many thanks

**Comment by igorsmotto** [comment]
Hi there @Fishbowler! Thanks for taking the time to create a Pull Request! Let me check internally with the team regarding your questions and I'll back to you.

Also, is there any technical reason why it was done for a partial set of commands? (Instead of all of them?)

**Comment by Fishbowler** [comment]
I did all the ones that seemed sensible. But I've only got my single point of view & use cases. I asked the question since it might go against some design values, rather than being "obviously incomplete".

**Comment by igorsmotto** [comment]
Hey @Fishbowler! Thanks for your answer - I was wondering if you had found any technical challenges but it seems it was not the case.

I brought this suggestion to the team, and we see this as a big quality-of-life improvement for Maestro users and **we want to move forward with this PR**. 

Now, one thing is that we would love to see this `label` for **all commands**. If you're up to implementing it for all the missing commands it would really be awesome, I can offer help with the review, merging, and release process of Maestro.

This also has the benefit of making it easier to update the documentation: we can make a new page that describes the label and that it works for every command. I can also help with documentation after this PR is ready for merge.

---
> Is there additional testing that could/should be done?

Your tests here are already great!

**Comment by Fishbowler** [comment]
@igorsmotto What are your feelings on the commands that currently don't take parameters, like `scroll` or `back`? Make this the first?

**Comment by igorsmotto** [comment]
I believe it makes sense. Since they will be optional.

The only thing to be aware of is that both APIs should work to avoid retro-compatibility issues, e.g.:
```
- scroll
```
or 
```
- scroll:
    label: "custom label"
```

**Comment by igorsmotto** [comment]
@Fishbowler 
I believe it makes sense. Since they will be optional.

The only thing to be aware of is that both APIs should work to avoid retro-compatibility issues, e.g.:
```
- scroll
```
and 
```
- scroll:
    label: "custom label"
```

**Comment by Fishbowler** [comment]
What about keeping support for undocumented (legacy?) formats like

```
- action: scroll
```

Currently, this is in my way.

**Comment by igorsmotto** [comment]
Feel free to ignore those.

**Comment by Fishbowler** [comment]
@igorsmotto I think this is ready for re-review

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Condition.kt [hunk 1]
```diff
@@ -20,6 +21,10 @@ data class Condition(
     }
 
     fun description(): String {
+        if(label != null && label != ""){
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Condition.kt:24]:
```suggestion
        if(label != null){
```

It's safe to assume that if the user decided to write the `label` that should be the description even if it's empty.

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 2]
```diff
@@ -51,11 +51,15 @@ data class SwipeCommand(
...
     val startRelative: String? = null,
     val endRelative: String? = null,
-    val duration: Long = DEFAULT_DURATION_IN_MILLIS
+    val duration: Long = DEFAULT_DURATION_IN_MILLIS,
+    val label: String? = null,
 ) : Command {
 
     override fun description(): String {
         return when {
+            label != null && label != "" -> {
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:60]:
```suggestion
            label != null -> {
```

See my comment on `Condition.kt`

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 3]
```diff
@@ -246,11 +261,12 @@ data class TapOnPointCommand(
...
     val waitUntilVisible: Boolean? = null,
     val longPress: Boolean? = null,
-    val repeat: TapRepeat? = null
+    val repeat: TapRepeat? = null,
+    val label: String? = null    
 ) : Command {
 
     override fun description(): String {
-        return "${tapOnDescription(longPress, repeat)} on point ($x, $y)"
+         return label ?: "${tapOnDescription(longPress, repeat)} on point ($x, $y)"
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:269]:
```suggestion
        return label ?: "${tapOnDescription(longPress, repeat)} on point ($x, $y)"
```

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 4]
```diff
@@ -282,9 +299,13 @@ data class AssertCommand(
     val visible: ElementSelector? = null,
     val notVisible: ElementSelector? = null,
     val timeout: Long? = null,
+    val label: String? = null,
 ) : Command {
 
     override fun description(): String {
+        if (label != null && label != ""){
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:306]:
```suggestion
        if (label != null){
```

See my comment on `Condition.kt`

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 5]
```diff
@@ -319,12 +340,13 @@ data class AssertCommand(
...
     val condition: Condition,
     private val timeout: String? = null,
+    val label: String? = null,
 ) : Command {
 
     fun timeoutMs() = timeout?.toLong()
 
     override fun description(): String {
-        return "Assert that ${condition.description()}"
+         return label ?: "Assert that ${condition.description()}"
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:349]:
```suggestion
        return label ?: "Assert that ${condition.description()}"
```

Style: Indent with 4 spaces

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 6]
```diff
@@ -336,11 +358,12 @@ data class AssertConditionCommand(
...
 
 data class InputTextCommand(
-    val text: String
+    val text: String,
+    val label: String? = null,
 ) : Command {
 
     override fun description(): String {
-        return "Input text $text"
+         return label ?: "Input text $text"
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:366]:
```suggestion
        return label ?: "Input text $text"
```

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 7]
```diff
@@ -357,9 +380,14 @@ data class LaunchAppCommand(
     val stopApp: Boolean? = null,
     var permissions: Map<String, String>? = null,
     val launchArguments: Map<String, Any>? = null,
+    val label: String? = null,
 ) : Command {
 
     override fun description(): String {
+        if (label != null && label != ""){
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:387]:
```suggestion
        if (label != null){
```

See my comment on `Condition.kt`

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 8]
```diff
@@ -412,11 +441,14 @@ data class ApplyConfigurationCommand(
...
     val link: String,
     val autoVerify: Boolean? = null,
-    val browser: Boolean? = null
+    val browser: Boolean? = null,
+    val label: String? = null,
 ) : Command {
 
     override fun description(): String {
-        return if (browser == true) {
+        return if (label != null && label != "") {
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:449]:
```suggestion
        return if (label != null) {
```

See my comment on `Condition.kt`

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 9]
```diff
@@ -445,10 +479,13 @@ data class PressKeyCommand(
 
 data class EraseTextCommand(
     val charactersToErase: Int?,
+    val label: String? = null,
 ) : Command {
 
     override fun description(): String {
-        return if (charactersToErase != null) {
+        return if (label != null && label != "") {
```
Comment (igorsmotto) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:486]:
```suggestion
        return if (label != null) {
```

See my comment on `Condition.kt`

File: maestro-test/src/test/kotlin/maestro/test/IntegrationTest.kt [hunk 10]
```diff
@@ -2475,7 +2474,7 @@ class IntegrationTest {
     }
 
     @Test
-    fun `091 - Assert visible by index`() {
+    fun `Case 091 - Assert visible by index`() {
```
Comment (igorsmotto) [inline maestro-test/src/test/kotlin/maestro/test/IntegrationTest.kt:2477]: Thank you! 👍

**Review by igorsmotto — COMMENTED** [review]
That's awesome work! @Fishbowler 🎉 

Made some small comments, but everything is looking good.

In the meantime, I'll run this PR against our internal test suite to be extra sure and if everything goes well, we'll be ready to merge it.

**Comment by Fishbowler** [comment]
Accepted all, and tidied commit history (slightly) 👍

**Comment by Fishbowler** [comment]
Added a draft PR of some docs to the docs repo. 

PR: https://github.com/mobile-dev-inc/maestro-docs/pull/25

Preview: https://github.com/Fishbowler/maestro-docs/blob/abe57ab66ab49c3c9361acaa130ec8ccc2290036/api-reference/labels.md

**Review by igorsmotto — APPROVED** [review]
Hi there, @Fishbowler!

After our internal testing, we're thrilled to report that everything has passed with flying colors. Therefore, your PR has been given the green light, and we're planning to merge it ASAP. 

I will take care of both this PR and the docs PR. 

One important note is that this change will wait until the next release (1.33.0). The exact release date is still TBD, but your name will be in the Changelog when it finally arrives. 

Your efforts have truly made a difference! We deeply appreciate it!

**Comment by igorsmotto** [comment]
Hey, @Fishbowler

Given that this is your user fork I don't have permission to edit this PR and resolve conflicts with the main branch

We have 3 options here:
1. Give me permission with `select Allow edits from maintainers.` [docs here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) so I can push the commit
2. Have a PR building with this work with the resolution of conflict [PR here](https://github.com/mobile-dev-inc/maestro/pull/1437). This will remove your author from the main branch history.
3. Cherry-pick this [commit](https://github.com/mobile-dev-inc/maestro/pull/1437/commits/f1f095641e78f847729da80012aa4fd651a59242) and update this branch so I can merge right away

**Comment by Fishbowler** [comment]
Weird, I've checked and I already had `Allow edits from maintainers` checked.

I've fast-forwarded my branch to yours, and got all the ✅s - good to go!

**Comment by Fishbowler** [comment]
Future archaeologists:
This was reverted after some internal problems in Maestro were discovered, then re-added by @amanjeetsingh150 in https://github.com/mobile-dev-inc/maestro/pull/1481

**Pull request merged**